### ISSUE TITLE
Why are we ignoring the changed file 

### ISSUE THREAD

**Issue by triHims** [GroovyLanguageServer/groovy-language-server#80]
https://github.com/GroovyLanguageServer/groovy-language-server/blob/4866a3f2c180f628405b1e4efbde0949a1418c10/src/main/java/net/prominic/groovyls/config/CompilationUnitFactory.java#L99C8-L99C8

Hi i am going though this codebase.
I see that we remove all the open files from Ast tree and then we try to get a compilationunit , We add every file to compilation unit expect open files in case where workspace root is not present. 
Why is that ? 
Thank you for reading.

**Comment by triHims** [comment]
Hey thank you very much for the response.
Can you also describe how do we handle changed files ?
They do not seem to be added anywhere.
Thanks again.

**Comment by joshtynjala** [comment]
Changes to files are handled in _GroovyServices.java_, in the `didChange` and `didChangeWatchedFiles` methods.

`didChange` is sent by VSCode (or any other editor using the language server) when the file is changed by the user in the editor. `didChangeWatchedFiles` is sent when the file changes on the file system (including by external programs).

**Comment by triHims** [comment]
I want to add support for jumping to code definition on the nodes annotated by autowired. 
Let me know what should I do for it ? 
I figured I could check the definition node here and if it has a annotation node of of autowired I can simply change the destination URI to the offsetNode  > Field > type's URI. 
Around here 
https://github.com/GroovyLanguageServer/groovy-language-server/blob/4866a3f2c180f628405b1e4efbde0949a1418c10/src/main/java/net/prominic/groovyls/providers/DefinitionProvider.java#L58

**Comment by joshtynjala** [comment]
Maybe. I'm not familiar with autowired or spring framework. I don't even use Groovy. I just know my way around creating language servers, and my client needed one for Groovy, so I've been trying to do my best.

**Comment by triHims** [comment]
Thank you for hard work that you do.
Can you tell me where to learn more of lsp and compilers?

**Comment by joshtynjala** [comment]
Here's the specification for the language server protocol. It shows all of the requests (and their responses), and notifications that are supported.

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/

It's just the raw protocol sent over stdin/stdout or a socket. Generally, you'll use a library that exposes the protocol to your language of choice. The Groovy language server uses the lsp4j library:

https://github.com/eclipse-lsp4j/lsp4j

I'm not sure I have many resources to suggest about compilers. I learned most of what I know by diving into a compiler codebase for a language I was using (ActionScript 3.0, in this case) and trying to add features and fix bugs. I eventually made a language server using that compiler. Luckily, the Groovy compiler was similar enough to the AS3 compiler (and both were implemented in Java) that I could make a similar language server for both.

I seem to recall that folks have mentioned this project as a good high-level overview of compilers, though:

https://github.com/jamiebuilds/the-super-tiny-compiler
